ffi/framebuffer.lua

```diff
@@ -138,9 +138,20 @@ function fb:setViewport(viewport)
     or viewport.w < 0 or viewport.x+viewport.w > self.screen_size.w
     or viewport.h < 0 or viewport.y+viewport.h > self.screen_size.h
     then
-        error("bad viewport")
+        error("fb:setViewport() bad viewport")
     end
-    self.full_bb = self.bb
+    -- we might be switching the viewport on the fly, so don't overwrite self.full_bb
+    if not self.full_bb then
+        self.full_bb = self.bb
+    else
+        -- reset before applying new viewport
+        self.bb = self.bb:viewport(
```

Comment:
@poire-z I'm not sure if this is quite the right way to go, although a viewport doesn't duplicate any memory and such.

Reply:
Is there a real need to reset?
https://github.com/koreader/koreader-base/blob/0b4c48a8e3fb64024e8d7d91090744c6bb8eeb39/ffi/blitbuffer.lua#L1135-L1141
Can't you just call it (so, no `else`) on self.full_bb :  `self.bb = self.full_bb:viewport()` ?

Reply:
I tried, but that just keeps pushing it down.

Reply:
Pardon, that's a bit incomplete. But my test case is a top "status bar height" on the Android emulator. I'll give provide some screenshots.

Reply:
Okay, so I was preparing my screenshots and I guess the problem I was trying to work around was caused by something incorrect in front? Or maybe it was just too late last night… but I'd swear what you said was the first thing I tried. Seems to be working now though.

There is this remaining beauty flaw. I know there's a draw white thing called somewhere…
<img src=https://user-images.githubusercontent.com/202757/31860661-cc4ce26e-b71d-11e7-8c72-31e6775d264c.png width=30%>